### PR TITLE
Fix Double Items in Dual Wielding

### PR DESCRIPTION
**Pull request by Spandamn**
- Fix Double of the same item on a pokemon
- Fix two choice items on a pokemon

### CODE REVIEW

**Comment by KrisOnline** [comment]
Why is there `validateSet` and `onValidateSet`

File: config/formats.js [hunk 1]
```diff
@@ -378,6 +378,12 @@ exports.Formats = [
 			let validator = new this.constructor(Dex.getFormat(this.format.id, ['Ignore Illegal Abilities']));
 			return validator.validateSet(Object.assign({}, set, {ability: ''}), teamHas) || validator.validateSet(Object.assign({}, set, {ability: '', item: set.ability}, teamHas));
 	  },
+		onValidateSet: function (set) {
+			let item = this.getItem(set.item);
+			let item2 = this.getItem(set.ability);
+			if (item.id === item2.id) return [`You cannot have two of the same item on a Pokemon. (${set.name || set.species} has two of ${this.getItem(set.item).name})`];
```
Comment (urkerab) [inline config/formats.js]: Might as well use `item.name` here too?

**Comment by urkerab** [comment]
Thanks for getting to this @Spandamn.

> Why is there validateSet and onValidateSet

`onValidateSet` is easier to use if you want to add extra bans to a regular set which can't be expressed in a banlist; Mix and Mega for instance bans certain stones except on their original users, while Middle Cup uses it to ensure you only use evolved Pokémon that have an evolution. `validateSet` is called instead of the regular validator; Gods and Followers uses it to validate the God differently to the Followers; Inheritance uses it to validate against a different Pokémon's template. It's possible to perform all of the `onValidateSet` validation in `validateSet` but it might make the coding slightly easier in this case.

**Comment by MacChaeger** [comment]
Did you test this? This doesn't seem like it would work since `onValidateSet` is only called from within the default `validateSet` which Dual Wielding overrides with its own `validateSet` that calls the regular one twice, once with one item in the item slot, and once with the other item, with an ability of `''` each time. As far as I can tell, `onValidateSet` will never be called in a context where the ability slot contains an item.

**Comment by MacChaeger** [comment]
Yeah, all this actually does is ban a Pokemon from having an item in their ability slot and nothing in their normal item slot.

**Comment by Zarel** [comment]
In case urkerab's response was unclear:

`validateSet` - replaces regular validation

`onValidateSet` - happens in addition to regular validation

**Comment by MacChaeger** [comment]
Actually it also bans having nothing in either slot too, which is actually currently allowed in Dual Wielding, and probably should continue to be, though there’s not exactly a competitive reason to do so other than say Acrobatics Archeops.

**Comment by Spandamn** [comment]
Thanks for the review guys, I updated it accordingly and it should work now

File: config/formats.js [hunk 2]
```diff
@@ -375,15 +375,13 @@ exports.Formats = [
 		validateSet: function (set, teamHas) {
 			let dual = this.dex.getItem(set.ability);
 			if (!dual.exists) return this.validateSet(set, teamHas);
+			let item = this.dex.getItem(set.item);
 			let validator = new this.constructor(Dex.getFormat(this.format.id, ['Ignore Illegal Abilities']));
-			return validator.validateSet(Object.assign({}, set, {ability: ''}), teamHas) || validator.validateSet(Object.assign({}, set, {ability: '', item: set.ability}, teamHas));
+			let problems = validator.validateSet(Object.assign({}, set, {ability: ''}), teamHas) || validator.validateSet(Object.assign({}, set, {ability: '', item: set.ability}, teamHas)) || [];
+			if (dual.id === item.id) problems.push(`You cannot have two of the same item on a Pokemon. (${set.name || set.species} has two of ${item.name})`);
+			if (item.isChoice && dual.isChoice) problems.push(`You cannot have two choice items on a Pokemon. (${set.name || set.species} has ${item.name} and ${dual.name})`);
```
Comment (urkerab) [inline config/formats.js]: If you equip 2 Choice Scarves then would you get two errors? (Or is that intentional?)

Comment (Spandamn) [reply config/formats.js]: Intentional, so that it reminds the user of both the rules

**Comment by Spandamn** [comment]
@TheImmortal

**Pull request merged**